docs/source/schemas/dynamic_columns.rst

```diff
@@ -0,0 +1,151 @@
+===============
+Dynamic Columns
+===============
+
+.. note::
+
+    This document discusses AtlasDB's dynamic columns. This is a separate concept from Cassandra's dynamic columns;
+    in fact, AtlasDB's schemas in Cassandra always use two Cassandra dynamic columns, independent of whether
+    we are storing a fixed-columns or dynamic columns table! Throughout this document, where "dynamic columns" is
+    used unqualified, we refer to AtlasDB dynamic columns.
+
+Overview
+--------
+
+AtlasDB offers support for *dynamic columns* - in tables where these are enabled, rows that can have an arbitrary number
+of columns, each with a (typed) column name. A table with dynamic columns may be reasoned about as a
+``Map<RowKey, SortedMap<Tuple<Col1, Col2... ColN>, Value>>``. It may also be thought of as a collection of key-value
+pairs within a row. For example, a simple schema using dynamic columns may be defined as follows:
+
+.. code:: java
+
+    schema.addTableDefinition("todo", new TableDefinition() {{
+                rowName();
+                    rowComponent("person", ValueType.STRING);
+                dynamicColumns();
+                    columnComponent("taskSize", ValueType.VAR_LONG);
+                    columnComponent("monetaryCost", ValueType.VAR_LONG);
+                    value(ValueType.STRING);
+            }});
+
+Note that dynamic column components must be primitive ValueTypes which support partitioning and ordering.
+Also, key-value pairs in an individual row will be returned in *lexicographically* sorted order of the key.
+
+We call the person the *row key*, the (taskSize, monetaryCost) pair the *dynamic column key*, and the string (which is
+intended to be a description of the todo) the *value*.
+
+Query Capabilities
+------------------
+
+Dynamic Columns are useful for avoiding KVS-level range scans, especially in key-value services where range scans
+are expensive (like Cassandra - its caching optimisations are rendered ineffectual for range scans). For example,
+for the schema defined above:
+
+1. "Find John's smallest and cheapest todo" can be answered very efficiently.
+   Elements are stored in sorted order, so the very first key-value pair retrieved from the database is actually the
+   correct answer to this query.
```

Comment:
Could use a code example - it looks like @sixinli wants to do exactly this.

Reply:
Yep! Query 1 is in the "retrieval" section. 

I've also added an appendix with implementations of a few more queries, with some comments explaining why they work. I've tested all of these locally as well. 

Query 8 was a bit tricky, but I validated it with @hsaraogi (thanks!).